go/p2p/peermgmt/backup.go

```diff
@@ -0,0 +1,194 @@
+	// storeBucketName is the name of the bucket in which backup data is stored.
+	storeBucketName = "p2p/peer_manager"
+
+	// storePeerstoreKey is the bucket key under which peers from the peerstore are stored.
+	storePeerstoreKey = "peerstore"
+
+	// backupStartupDelay is the startup time delay for backups.
+	backupStartupDelay = 15 * time.Minute
+
+	// backupInterval is the time interval between backups.
+	backupInterval = 15 * time.Minute
+
+	// maxRestorePeers is the maximum number of peers connected on startup from the backup.
+	maxRestorePeers = 100
+)
+
+type peerstoreBackup struct {
+	logger *logging.Logger
+
+	host host.Host
+
+	store   *persistent.CommonStore
+	storeCh chan struct{}
+
+	runCh chan struct{}
+}
+
+func newPeerstoreBackup(h host.Host, cs *persistent.CommonStore) *peerstoreBackup {
+	l := logging.GetLogger("p2p/peer-manager/backup")
+
+	return &peerstoreBackup{
+		logger:  l,
+		host:    h,
+		store:   cs,
+		storeCh: make(chan struct{}, 1),
+		runCh:   make(chan struct{}, 1),
+	}
+}
+
+// save persists peers to the store.
+func (b *peerstoreBackup) save() error {
+	if b.store == nil {
+		return nil
+	}
+
+	// Allow only one save/load at a time.
+	b.storeCh <- struct{}{}
+	defer func() { <-b.storeCh }()
+
+	// Save only connected peers.
+	peers := b.host.Network().Peers()
+
+	if len(peers) == 0 {
+		return nil
+	}
+
+	b.logger.Debug("storing peers",
+		"count", len(peers),
+	)
+
+	infos := make([][]byte, 0, len(peers))
+	for _, p := range peers {
+		info := b.host.Peerstore().PeerInfo(p)
+		if len(info.Addrs) == 0 {
+			continue
+		}
+
+		json, err := info.MarshalJSON()
+		if err != nil {
+			return err
+		}
+
+		infos = append(infos, json)
+	}
+
+	// Persist addresses.
+	bucket, err := b.store.GetServiceStore(storeBucketName)
+	if err != nil {
+		return err
+	}
+	if err = bucket.PutCBOR([]byte(storePeerstoreKey), infos); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// load loads peers from the store.
+func (b *peerstoreBackup) load() ([]*peer.AddrInfo, error) {
+	if b.store == nil {
+		return []*peer.AddrInfo{}, nil
+	}
+
+	// Allow only one save/load at a time.
+	b.storeCh <- struct{}{}
+	defer func() { <-b.storeCh }()
+
+	b.logger.Debug("loading peers")
+
+	// Load addresses in json form.
+	bucket, err := b.store.GetServiceStore(storeBucketName)
+	if err != nil {
+		return nil, err
+	}
+	var jsons [][]byte
+	if err = bucket.GetCBOR([]byte(storePeerstoreKey), &jsons); err != nil {
+		return nil, err
+	}
+
+	// Convert them.
+	infos := make([]*peer.AddrInfo, len(jsons))
+	for i, json := range jsons {
+		var info peer.AddrInfo
+		if err = info.UnmarshalJSON(json); err != nil {
+			return nil, err
+		}
+		infos[i] = &info
+	}
+
+	return infos, nil
+}
+
+// run periodically backups peers to the store. If a backup is already in progress, this method will
+// wait for it to finish.
+func (b *peerstoreBackup) run(ctx context.Context) {
+	// Allow to run only one backup at a time.
+	select {
+	case b.runCh <- struct{}{}:
```

Comment:
This seems a bit weird given that `run` does not seem to return until context is cancelled which only happens during node shutdown? I think it makes more sense to make this a no-op if already running?

Also it seems weird to make this based on a context as this seems more like a service with a start/stop?

Reply:
Changed to no-op if already running and switched to start/stop approach.